PVTPackage/source/MultiphaseSystem/DeadOilMultiphaseSystem.cpp

```diff
@@ -62,15 +92,43 @@ std::unique_ptr< DeadOilMultiphaseSystem > DeadOilMultiphaseSystem::build( const
                                                                            const std::vector< double > & molarWeights )
 {
+  const bool containsOil = std::find( phases.cbegin(), phases.cend(), pvt::PHASE_TYPE::OIL ) != phases.cend();
+  const bool containsGas = std::find( phases.cbegin(), phases.cend(), pvt::PHASE_TYPE::GAS ) != phases.cend();
+  const bool containsWater = std::find( phases.cbegin(), phases.cend(), pvt::PHASE_TYPE::LIQUID_WATER_RICH ) != phases.cend();  
+
+  const bool isValid = containsOil && ( containsGas || containsWater );
+  if( !isValid )  
+  {    
+    const std::string msg = "Three types of DO systems are allowed: Oil-Water-Gas, Oil-Water, and Oil-Gas";
+    LOGERROR( msg );
+  }
+   
   // props.oilTable, props.gasTable and props.waterTable respectively contain PVDO, PVDG and PVTW
   const Properties & props = buildTables( phases, tableFileNames, surfaceDensities, molarWeights );
 
   // I am not using std::make_unique because I want the constructor to be private.
-  auto * ptr = new DeadOilMultiphaseSystem( phases,
-                                            props.oilTable, props.oilSurfaceMassDensity, props.oilSurfaceMolecularWeight,
-                                            props.gasTable, props.gasSurfaceMassDensity, props.gasSurfaceMolecularWeight,
-                                            props.waterTable, props.waterSurfaceMassDensity, props.waterSurfaceMolecularWeight );
-  return std::unique_ptr< DeadOilMultiphaseSystem >( ptr );
+  if( containsOil && containsGas && containsWater )
+  {    
+    auto * ptr = new DeadOilMultiphaseSystem( phases,
+                                              props.oilTable, props.oilSurfaceMassDensity, props.oilSurfaceMolecularWeight,
+                                              props.gasTable, props.gasSurfaceMassDensity, props.gasSurfaceMolecularWeight,
+                                              props.waterTable, props.waterSurfaceMassDensity, props.waterSurfaceMolecularWeight );
+    return std::unique_ptr< DeadOilMultiphaseSystem >( ptr );
+  }
+  else if( containsOil && containsGas )
+  {
+    auto * ptr = new DeadOilMultiphaseSystem( phases,
+                                              props.oilTable, props.oilSurfaceMassDensity, props.oilSurfaceMolecularWeight,
+                                              props.gasTable, props.gasSurfaceMassDensity, props.gasSurfaceMolecularWeight );
+    return std::unique_ptr< DeadOilMultiphaseSystem >( ptr );
+  }
+  else // containsOil && containsWater
+  {
+    auto * ptr = new DeadOilMultiphaseSystem( phases,
+                                              props.oilTable, props.oilSurfaceMassDensity, props.oilSurfaceMolecularWeight,
+                                              props.waterTable, props.waterSurfaceMassDensity, props.waterSurfaceMolecularWeight );
+    return std::unique_ptr< DeadOilMultiphaseSystem >( ptr );    
+  }
```

Comment:
Something like this could be even simpler, what do you think?
The flow would seem more natural, fluid.

```suggestion
  else if( containsOil && containsWater )
  {
    auto * ptr = new DeadOilMultiphaseSystem( phases,
                                              props.oilTable, props.oilSurfaceMassDensity, props.oilSurfaceMolecularWeight,
                                              props.waterTable, props.waterSurfaceMassDensity, props.waterSurfaceMolecularWeight );
    return std::unique_ptr< DeadOilMultiphaseSystem >( ptr );    
  }
  else
  {
    const std::string msg = "Three types of DO systems are allowed: Oil-Water-Gas, Oil-Water, and Oil-Gas";
    LOGERROR( msg );
    // Image we do not want to kill the process, we could return a null unique_ptr (bad + doc!) so the client knows ;)
  }
```

Reply:
Done